In [ ]:
%pip install mlflow

Python interpreter will be restarted.
Collecting mlflow
 Using cached mlflow-1.29.0-py3-none-any.whl (16.9 MB)
Collecting click<9,>=7.0
 Using cached click-8.1.3-py3-none-any.whl (96 kB)
Collecting alembic<2
 Using cached alembic-1.8.1-py3-none-any.whl (209 kB)
Requirement already satisfied: pandas<2 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.0.1)
Collecting querystring-parser<2
 Using cached querystring_parser-1.2.4-py2.py3-none-any.whl (7.9 kB)
Requirement already satisfied: protobuf<5,>=3.12.0 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (3.20.1)
Collecting importlib-metadata!=4.7.0,<5,>=3.7.0
 Using cached importlib_metadata-4.13.0-py3-none-any.whl (23 kB)
Collecting sqlparse<1,>=0.4.0
 Using cached sqlparse-0.4.3-py3-none-any.whl (42 kB)
Requirement already satisfied: numpy<2 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.18.1)
Collecting sqlalchemy<2,>=1.4.0
 Using cached SQLAlchemy-1.4.41-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.6 MB)
Collecting docker<7,>=4.0.0
 Using cached docker-6.0.0-py3-none-any.whl (147 kB)
Requirement already satisfied: requests<3,>=2.17.3 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (2.22.0)
Collecting gitpython<4,>=2.1.0
 Using cached GitPython-3.1.27-py3-none-any.whl (181 kB)
Collecting gunicorn<21; platform_system != "Windows"
 Using cached gunicorn-20.1.0-py3-none-any.whl (79 kB)
Requirement already satisfied: scipy<2 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.4.1)
Collecting cloudpickle<3
 Using cached cloudpickle-2.2.0-py3-none-any.whl (25 kB)
Collecting packaging<22
 Using cached packaging-21.3-py3-none-any.whl (40 kB)
Requirement already satisfied: entrypoints<1 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (0.3)
Processing /root/.cache/pip/wheels/3f/73/87/c1e4b2145eb6049bb6c9aaf7ea1e38302b77ca219b6fef5d5c/databricks_cli-0.17.3-py3-none-any.whl
Requirement already satisfied: pytz<2023 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (2019.3)
Collecting Flask<3
 Using cached Flask-2.2.2-py3-none-any.whl (101 kB)
Collecting pyyaml<7,>=5.1
 Using cached PyYAML-6.0-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (596 kB)
Collecting prometheus-flask-exporter<1
 Using cached prometheus_flask_exporter-0.20.3-py3-none-any.whl (18 kB)
Collecting Mako
 Using cached Mako-1.2.3-py3-none-any.whl (78 kB)
Collecting importlib-resources; python_version < "3.9"
 Using cached importlib_resources-5.9.0-py3-none-any.whl (33 kB)
Requirement already satisfied: python-dateutil>=2.6.1 in /databricks/python3/lib/python3.7/site-packages (from pandas<2->mlflow) (2.8.1)
Requirement already satisfied: six in /databricks/python3/lib/python3.7/site-packages (from querystring-parser<2->mlflow) (1.14.0)
Collecting typing-extensions>=3.6.4; python_version < "3.8"
 Using cached typing_extensions-4.3.0-py3-none-any.whl (25 kB)
Collecting zipp>=0.5
 Using cached zipp-3.8.1-py3-none-any.whl (5.6 kB)
Collecting greenlet!=0.4.17; python_version >= "3" and (platform_machine == "aarch64" or (platform_machine == "ppc64le" or (platform_machine == "x86_64" or (platform_machine == "amd64" or (platform_machine == "AMD64" or (platform_machine == "win32" or platform_machine == "WIN32"))))))
 Using cached greenlet-1.1.3-cp37-cp37m-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (150 kB)
Collecting websocket-client>=0.32.0
 Using cached websocket_client-1.4.1-py3-none-any.whl (55 kB)
Collecting urllib3>=1.26.0
 Using cached urllib3-1.26.12-py2.py3-none-any.whl (140 kB)
Requirement already satisfied: idna<2.9,>=2.5 in /databricks/python3/lib/python3.7/site-packages (from requests<3,>=2.17.3->mlflow) (2.8)
Requirement already satisfied: certifi>=2017.4.17 in /databricks/python3/lib/python3.7/site-packages (from requests<3,>=2.17.3->mlflow) (2020.6.20)
Requirement already satisfied: chardet<

In [ ]:
#all spark imports
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql.functions import *
from pyspark.sql.functions import desc

from pyspark.ml.recommendation import ALS
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.feature import Imputer, StringIndexer, VectorAssembler
from pyspark.ml import Pipeline
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder

In [ ]:
#instantiate the spark session
spark = SparkSession.builder.appName("Demo").getOrCreate()

## Part A

### 1. Count the odd and even numbers using file ‘integer.txt’

In [ ]:
integerRDD = spark.sparkContext.textFile("/FileStore/tables/integer.txt")

In [ ]:
#Number of whole integers
integerRDD.count()

Out[4]: 1010

In [ ]:
#create rdds for odd numbers and even numbers respectively
odd_RDD = integerRDD.filter(lambda x: int(x) % 2 == 0)
even_RDD = integerRDD.filter(lambda x: int(x) % 2 != 0)

In [ ]:
#Show the results
print("Number of odd integers:",odd_RDD.count(),",","number of even integers:",even_RDD.count())

Number of odd integers: 514 , number of even integers: 496

### 2. Calculate the salary sum per department using file ‘salary.txt’. Show department name and salary sum.

In [ ]:
salaryRDD = spark.sparkContext.textFile("/FileStore/tables/salary.txt")

In [ ]:
#split each line by ","
arrayRDD = salaryRDD.map(lambda x: x.split(","))
#split name and salary by " "
arrayRDD = arrayRDD.map(lambda x: x[0].split(" "))
#make pair of each name and its salary 
kvRDD = arrayRDD.map(lambda x: (x[0],int(x[1])))
#sum by key(department name)
sumRDD = kvRDD.reduceByKey(lambda x,y: x+y)

In [ ]:
#Show department name and salary sum
print(sumRDD.collect())

[('Sales', 3488491), ('Research', 3328284), ('Developer', 3221394), ('QA', 3360624), ('Marketing', 3158450)]

### 3. Implement MapReduce using Pyspark on file ‘shakespeare.txt’. Show how many times these particular words appear in the document: Shakespeare, why, Lord, Library, GUTENBERG, WILLIAM, COLLEGE and WORLD. (Count exact words only (marks will be deducted for incorrect lowercase/uppercase))

In [ ]:
shakespeareRDD = spark.sparkContext.textFile("/FileStore/tables/shakespeare.txt")

In [ ]:
#function to remore punctuation
import re
def removePunctuation(text):
    return re.sub(re.compile(r'[^a-zA-Z0-9\s]'),"",text).strip()

In [ ]:
#remove all punctuation in shakespeaseRDD
cleanRDD = shakespeareRDD.map(removePunctuation)

In [ ]:
cleanRDD.collect()

Out[13]: ['This eBook is for the use of anyone anywhere at no cost and with',
 'almost no restrictions whatsoever You may copy it give it away or',
 'reuse it under the terms of the Project Gutenberg License included',
 'with this eBook or online at wwwgutenbergorg',
 '',
 'This is a COPYRIGHTED Project Gutenberg eBook Details Below',
 'Please follow the copyright guidelines in this file',
 '',
 'Title The Complete Works of William Shakespeare',
 '',
 'Author William Shakespeare',
 '',
 'Posting Date September 1 2011 EBook 100',
 'Release Date January 1994',
 '',
 'Language English',
 '',
 'Character set encoding ASCII',
 '',
 'START OF THIS PROJECT GUTENBERG EBOOK COMPLETE WORKSWILLIAM SHAKESPEARE',
 '',
 '',
 '',
 '',
 'Produced by World Library Inc from their Library of the Future',
 '',
 '',
 '',
 '',
 'This is the 100th Etext file presented by Project Gutenberg and',
 'is presented in cooperation with World Library Inc from their',
 'Library of the Future and Shakespeare CDROMS Project Gutenberg',
 'often releases Etexts that are NOT placed in the Public Domain',
 '',
 'Shakespeare',
 '',
 'This Etext has certain copyright implications you should read',
 '',
 'THIS ELECTRONIC VERSION OF THE COMPLETE WORKS OF WILLIAM',
 'SHAKESPEARE IS COPYRIGHT 19901993 BY WORLD LIBRARY INC AND IS',
 'PROVIDED BY PROJECT GUTENBERG ETEXT OF ILLINOIS BENEDICTINE COLLEGE',
 'WITH PERMISSION ELECTRONIC AND MACHINE READABLE COPIES MAY BE',
 'DISTRIBUTED SO LONG AS SUCH COPIES 1 ARE FOR YOUR OR OTHERS',
 'PERSONAL USE ONLY AND 2 ARE NOT DISTRIBUTED OR USED',
 'COMMERCIALLY PROHIBITED COMMERCIAL DISTRIBUTION INCLUDES BY ANY',
 'SERVICE THAT CHARGES FOR DOWNLOAD TIME OR FOR MEMBERSHIP',
 '',
 'Project Gutenberg is proud to cooperate with The World Library',
 'in the presentation of The Complete Works of William Shakespeare',
 'for your reading for education and entertainment HOWEVER THIS',
 'IS NEITHER SHAREWARE NOR PUBLIC DOMAIN AND UNDER THE LIBRARY',
 'OF THE FUTURE CONDITIONS OF THIS PRESENTATION NO CHARGES MAY',
 'BE MADE FOR ANY ACCESS TO THIS MATERIAL YOU ARE ENCOURAGED',
 'TO GIVE IT AWAY TO ANYONE YOU LIKE BUT NO CHARGES ARE ALLOWED',
 '',
 '',
 '',
 '',
 'SMALL PRINT for COMPLETE SHAKESPEARE',
 '',
 'THIS ELECTRONIC VERSION OF THE COMPLETE WORKS OF WILLIAM',
 'SHAKESPEARE IS COPYRIGHT 19901993 BY WORLD LIBRARY INC',
 'AND IS PROVIDED BY PROJECT GUTENBERG ETEXT OF',
 'ILLINOIS BENEDICTINE COLLEGE WITH PERMISSION',
 '',
 'Since unlike many other Project Gutenbergtm etexts this etext',
 'is copyright protected and since the materials and methods you',
 'use will effect the Projects reputation your right to copy and',
 'distribute it is limited by the copyright and other laws and by',
 'the conditions of this Small Print statement',
 '',
 '1 LICENSE',
 '',
 'A YOU MAY AND ARE ENCOURAGED TO DISTRIBUTE ELECTRONIC AND',
 'MACHINE READABLE COPIES OF THIS ETEXT SO LONG AS SUCH COPIES',
 '1 ARE FOR YOUR OR OTHERS PERSONAL USE ONLY AND 2 ARE NOT',
 'DISTRIBUTED OR USED COMMERCIALLY PROHIBITED COMMERCIAL',
 'DISTRIBUTION INCLUDES BY ANY SERVICE THAT CHARGES FOR DOWNLOAD',
 'TIME OR FOR MEMBERSHIP',
 '',
 'B This license is subject to the conditions that you honor',
 'the refund and replacement provisions of this small print',
 'statement and that you distribute exact copies of this etext',
 'including this Small Print statement Such copies can be',
 'compressed or any proprietary form including any form resulting',
 'from word processing or hypertext software so long as',
 'EITHER',
 '',
 '1 The etext when displayed is clearly readable and does',
 'not contain characters other than those intended by the',
 'author of the work although tilde asterisk and',
 'underline characters may be used to convey punctuation',
 'intended by the author and additional characters may be used',
 'to indicate hypertext links OR',
 '',
 '2 The etext is readily convertible by the reader at no',
 'expense into plain ASCII EBCDIC or equivalent form by the',
 'program tha

In [ ]:
#split into each word
wordRDD = cleanRDD.flatMap(lambda x: x.split())

In [ ]:
wordRDD.collect()

Out[15]: ['This',
 'eBook',
 'is',
 'for',
 'the',
 'use',
 'of',
 'anyone',
 'anywhere',
 'at',
 'no',
 'cost',
 'and',
 'with',
 'almost',
 'no',
 'restrictions',
 'whatsoever',
 'You',
 'may',
 'copy',
 'it',
 'give',
 'it',
 'away',
 'or',
 'reuse',
 'it',
 'under',
 'the',
 'terms',
 'of',
 'the',
 'Project',
 'Gutenberg',
 'License',
 'included',
 'with',
 'this',
 'eBook',
 'or',
 'online',
 'at',
 'wwwgutenbergorg',
 'This',
 'is',
 'a',
 'COPYRIGHTED',
 'Project',
 'Gutenberg',
 'eBook',
 'Details',
 'Below',
 'Please',
 'follow',
 'the',
 'copyright',
 'guidelines',
 'in',
 'this',
 'file',
 'Title',
 'The',
 'Complete',
 'Works',
 'of',
 'William',
 'Shakespeare',
 'Author',
 'William',
 'Shakespeare',
 'Posting',
 'Date',
 'September',
 '1',
 '2011',
 'EBook',
 '100',
 'Release',
 'Date',
 'January',
 '1994',
 'Language',
 'English',
 'Character',
 'set',
 'encoding',
 'ASCII',
 'START',
 'OF',
 'THIS',
 'PROJECT',
 'GUTENBERG',
 'EBOOK',
 'COMPLETE',
 'WORKSWILLIAM',
 'SHAKESPEARE',
 'Produced',
 'by',
 'World',
 'Library',
 'Inc',
 'from',
 'their',
 'Library',
 'of',
 'the',
 'Future',
 'This',
 'is',
 'the',
 '100th',
 'Etext',
 'file',
 'presented',
 'by',
 'Project',
 'Gutenberg',
 'and',
 'is',
 'presented',
 'in',
 'cooperation',
 'with',
 'World',
 'Library',
 'Inc',
 'from',
 'their',
 'Library',
 'of',
 'the',
 'Future',
 'and',
 'Shakespeare',
 'CDROMS',
 'Project',
 'Gutenberg',
 'often',
 'releases',
 'Etexts',
 'that',
 'are',
 'NOT',
 'placed',
 'in',
 'the',
 'Public',
 'Domain',
 'Shakespeare',
 'This',
 'Etext',
 'has',
 'certain',
 'copyright',
 'implications',
 'you',
 'should',
 'read',
 'THIS',
 'ELECTRONIC',
 'VERSION',
 'OF',
 'THE',
 'COMPLETE',
 'WORKS',
 'OF',
 'WILLIAM',
 'SHAKESPEARE',
 'IS',
 'COPYRIGHT',
 '19901993',
 'BY',
 'WORLD',
 'LIBRARY',
 'INC',
 'AND',
 'IS',
 'PROVIDED',
 'BY',
 'PROJECT',
 'GUTENBERG',
 'ETEXT',
 'OF',
 'ILLINOIS',
 'BENEDICTINE',
 'COLLEGE',
 'WITH',
 'PERMISSION',
 'ELECTRONIC',
 'AND',
 'MACHINE',
 'READABLE',
 'COPIES',
 'MAY',
 'BE',
 'DISTRIBUTED',
 'SO',
 'LONG',
 'AS',
 'SUCH',
 'COPIES',
 '1',
 'ARE',
 'FOR',
 'YOUR',
 'OR',
 'OTHERS',
 'PERSONAL',
 'USE',
 'ONLY',
 'AND',
 '2',
 'ARE',
 'NOT',
 'DISTRIBUTED',
 'OR',
 'USED',
 'COMMERCIALLY',
 'PROHIBITED',
 'COMMERCIAL',
 'DISTRIBUTION',
 'INCLUDES',
 'BY',
 'ANY',
 'SERVICE',
 'THAT',
 'CHARGES',
 'FOR',
 'DOWNLOAD',
 'TIME',
 'OR',
 'FOR',
 'MEMBERSHIP',
 'Project',
 'Gutenberg',
 'is',
 'proud',
 'to',
 'cooperate',
 'with',
 'The',
 'World',
 'Library',
 'in',
 'the',
 'presentation',
 'of',
 'The',
 'Complete',
 'Works',
 'of',
 'William',
 'Shakespeare',
 'for',
 'your',
 'reading',
 'for',
 'education',
 'and',
 'entertainment',
 'HOWEVER',
 'THIS',
 'IS',
 'NEITHER',
 'SHAREWARE',
 'NOR',
 'PUBLIC',
 'DOMAIN',
 'AND',
 'UNDER',
 'THE',
 'LIBRARY',
 'OF',
 'THE',
 'FUTURE',
 'CONDITIONS',
 'OF',
 'THIS',
 'PRESENTATION',
 'NO',
 'CHARGES',
 'MAY',
 'BE',
 'MADE',
 'FOR',
 'ANY',
 'ACCESS',
 'TO',
 'THIS',
 'MATERIAL',
 'YOU',
 'ARE',
 'ENCOURAGED',
 'TO',
 'GIVE',
 'IT',
 'AWAY',
 'TO',
 'ANYONE',
 'YOU',
 'LIKE',
 'BUT',
 'NO',
 'CHARGES',
 'ARE',
 'ALLOWED',
 'SMALL',
 'PRINT',
 'for',
 'COMPLETE',
 'SHAKESPEARE',
 'THIS',
 'ELECTRONIC',
 'VERSION',
 'OF',
 'THE',
 'COMPLETE',
 'WORKS',
 'OF',
 'WILLIAM',
 'SHAKESPEARE',
 'IS',
 'COPYRIGHT',
 '19901993',
 'BY',
 'WORLD',
 'LIBRARY',
 'INC',
 'AND',
 'IS',
 'PROVIDED',
 'BY',
 'PROJECT',
 'GUTENBERG',
 'ETEXT',
 'OF',
 'ILLINOIS',
 'BENEDICTINE',
 'COLLEGE',
 'WITH',
 'PERMISSION',
 'Since',
 'unlike',
 'many',
 'other',
 'Project',
 'Gutenbergtm',
 'etexts',
 'this',
 'etext',
 'is',
 'copyright',
 'protected',
 'and',
 'since',
 'the',
 'materials',
 'and',
 'methods',
 'you',
 'use',
 'will',
 'effect',
 'the',
 'Projects',
 'reputation',
 'your',
 'right',
 'to',
 'copy',
 'and',
 'distribute',
 'it',
 'is',
 'limited',
 'by',
 'the',
 'copyright',
 'and',
 'other',
 'laws',
 'and',
 'by',
 'the',
 'conditions',
 'of',
 'this',
 'Smal

In [ ]:
#List the exact words
word_list = ['Shakespeare', 'why', 'Lord', 'Library', 'GUTENBERG', 'WILLIAM', 'COLLEGE', 'WORLD']

In [ ]:
#create dictionary to save the counts for each word
count_words = {}
for word in word_list:
    count_words[word] = wordRDD.filter(lambda x: x == word).count()

In [ ]:
#Show how many times these particular words appear in the document: Shakespeare, why, Lord, Library, GUTENBERG, WILLIAM, COLLEGE and WORLD.
print(count_words)

{'Shakespeare': 22, 'why': 114, 'Lord': 400, 'Library': 5, 'GUTENBERG': 99, 'WILLIAM': 127, 'COLLEGE': 98, 'WORLD': 98}

### 4. Calculate top 10 and bottom 10 words using file ‘shakespeare.txt’. Show 10 words with most count and 10 words with least count. You can limit by 10 in ascending and descending order of count. Show your code and output

In [ ]:
#create pair for each word (word,1)
kvRDD = wordRDD.map(lambda x: (x,1))

In [ ]:
#sum the value by key
sumRDD = kvRDD.reduceByKey(lambda x,y: x+y)

In [ ]:
#Top 10 words: Sort by value descending and show first 10 items
top10 = sumRDD.sortBy(lambda x: x[1],False).collect()[:10]
print(top10)

[('the', 11402), ('I', 8936), ('and', 8916), ('of', 7957), ('to', 7508), ('a', 5755), ('you', 5306), ('my', 4919), ('in', 4722), ('that', 3754)]

In [ ]:
#Bottom 10 words: Sort by value ascending and show first 10 items
bottom10 = sumRDD.sortBy(lambda x: x[1],True).collect()[:10]
print(bottom10)

[('anyone', 1), ('restrictions', 1), ('reuse', 1), ('online', 1), ('COPYRIGHTED', 1), ('Details', 1), ('guidelines', 1), ('Posting', 1), ('2011', 1), ('EBook', 1)]

## Part B

### 1. Describe your data. Calculate top 20 movies with highest ratings and top 15 users who provided highest ratings. Show your code and output.

In [ ]:
# File location and type
path = "/FileStore/tables/movies.csv"
 
df = spark.read \
  .format("csv") \
  .option("inferSchema", True) \
  .option("header", True) \
  .option("sep", ',') \
  .option("path", path) \
  .load()

In [ ]:
df.rdd.getNumPartitions()

Out[5]: 1

In [ ]:
df.printSchema()

root
-- movieId: integer (nullable = true)
-- rating: integer (nullable = true)
-- userId: integer (nullable = true)

In [ ]:
df.describe().show()

+-------+------------------+------------------+------------------+
summary| movieId| rating| userId|
+-------+------------------+------------------+------------------+
 count| 1501| 1501| 1501|
 mean| 49.40572951365756|1.7741505662891406|14.383744170552964|
 stddev|28.937034065088994| 1.187276166124803| 8.591040424293272|
 min| 0| 1| 0|
 max| 99| 5| 29|
+-------+------------------+------------------+------------------+

In this data, there are 1501 ratings from 30 users for 100 movies. There are three columns: movieID, rating and userID. These three columns are all integers. The rating values is from 1 to 5.

In [ ]:
#Top 20 movies(groupby 'movieId' and sort by count)
df.groupby('movieId').count().sort(desc('count')).show(15)

+-------+-----+
movieId|count|
+-------+-----+
 6| 20|
 29| 20|
 51| 20|
 22| 20|
 50| 20|
 94| 19|
 55| 19|
 68| 19|
 2| 19|
 15| 19|
 85| 18|
 36| 18|
 86| 18|
 88| 18|
 45| 18|
+-------+-----+
only showing top 15 rows

In [ ]:
#Top 15 users who provided highest ratings(groupby 'userId' and sort by count)
df.groupBy('userId').count().sort(desc('count')).show(15)

+------+-----+
userId|count|
+------+-----+
 6| 57|
 14| 57|
 22| 56|
 11| 56|
 12| 55|
 4| 55|
 7| 54|
 9| 53|
 24| 52|
 18| 52|
 23| 52|
 28| 50|
 26| 49|
 19| 49|
 1| 49|
+------+-----+
only showing top 15 rows

### 2. Split dataset into train and test. Try 2 different combinations (60/40, 80/20). Show your code and output.

In [ ]:
#Try 1: ramdom split into 60/40
train_1,test_1 = df.randomSplit([0.6, 0.4],seed=2022)
train_1.count(),test_1.count()

Out[52]: (905, 596)

In [ ]:
# training the model
als = ALS(maxIter=5, regParam=0.01, userCol="userId", itemCol="movieId", ratingCol="rating", coldStartStrategy="drop")
model_1 = als.fit(train_1)

In [ ]:
# predict using the testing datatset
predictions_1 = model_1.transform(test_1)
predictions_1.show()

+-------+------+------+------------+
movieId|rating|userId| prediction|
+-------+------+------+------------+
 31| 1| 4| 3.2615635|
 31| 1| 5| 0.6294935|
 31| 3| 7| 1.9749813|
 31| 4| 12| 1.311902|
 31| 1| 19| 1.2478268|
 31| 1| 29| 1.9299847|
 85| 3| 1| 1.9656754|
 85| 1| 5| 1.3060498|
 85| 3| 6| 2.2567053|
 85| 5| 16| 0.79389894|
 85| 2| 20| 1.1188357|
 85| 3| 21| -1.0086721|
 85| 1| 25| -1.5863776|
 85| 1| 29| 1.2430565|
 65| 2| 3| 2.7134957|
 65| 1| 4| 1.3937733|
 65| 1| 22|-0.062134743|
 65| 1| 28| 1.8804145|
 53| 5| 8| 1.9821519|
 53| 1| 9| 1.4304165|
+-------+------+------+------------+
only showing top 20 rows

In [ ]:
#Try 2: ramdom split into 80/20
train_2,test_2 = df.randomSplit([0.8, 0.2],seed=2022)
train_2.count(),test_2.count()

Out[48]: (1210, 291)

In [ ]:
# training the model
als = ALS(maxIter=5, regParam=0.01, userCol="userId", itemCol="movieId", ratingCol="rating", coldStartStrategy="drop")
model_2 = als.fit(train_2)

In [ ]:
# predict using the testing datatset
predictions_2 = model_2.transform(test_2)
predictions_2.show()

+-------+------+------+-----------+
movieId|rating|userId| prediction|
+-------+------+------+-----------+
 31| 1| 5| 1.1093016|
 31| 4| 12| 0.11865923|
 85| 1| 5|-0.37261903|
 85| 3| 6| 1.7296251|
 85| 5| 16| 1.9365113|
 85| 1| 25| 1.6636505|
 65| 2| 3| 2.571944|
 65| 1| 28| -1.3823442|
 53| 1| 9| 2.3018074|
 53| 3| 14| 3.687815|
 53| 2| 19|-0.07052807|
 53| 3| 20|-0.12884046|
 53| 1| 25| 0.33389556|
 78| 1| 19| 0.9529345|
 34| 1| 4| 1.1451671|
 34| 1| 16| 0.57677126|
 34| 1| 17| 0.98063934|
 81| 1| 1| -2.5154011|
 81| 1| 3| 2.070742|
 81| 2| 5| 2.1712854|
+-------+------+------+-----------+
only showing top 20 rows

### 3. Explain MSE, RMSE and MAE. Compare and evaluate both of your models with evaluation metrics (RMSE), show your code and print your results. Describe which one works better and why?

1. Mean absolute error (MAE) represents the average of the absolute differences between the actual and predicted values in the data set, and it measures the average of the residuals in the data set.
2. Mean squared error (MSE) represents the average of the squared differences between the original and predicted values in the data set. It measures the variance of the residuals.
3. The root mean squared error (RMSE) is the square root of the mean squared error (MSE). It measures the standard deviation of the residuals.

MSE and RMSE penalize larger prediction errors by MAE. However, RMSE is more widely used than MSE to assess the performance of regression models versus other stochastic models because it has the same units as the dependent variable (Y-axis).

In [ ]:
# Evaluate the model 1 (60/40) by computing the RMSE on the test data
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
rmse_1 = evaluator.evaluate(predictions_1)
print("Root-mean-square error = " + str(rmse_1))

Root-mean-square error = 2.0452337040005486

In [ ]:
#Evaluate the model 2 (80/20) by computing the RMSE on the test data
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
rmse_2 = evaluator.evaluate(predictions_2)
print("Root-mean-square error = " + str(rmse_2))

Root-mean-square error = 1.9820822053851181

The lower value of RMSE means the higher accuracy of the model. Thus, in these two cases, spliting training/test into 80/20 for model 2 has better performance. Increasing the percentage of the training set can make the model prediction more accurate.

### 4. Now tune the parameters of your algorithm to get the best set of parameters. Explain different parameters of the algorithm which you have used for tuning your algorithm. Evaluate all your models again. Show your code with best values and output.

The four parameters I used to tune the model is: rank, maxIter, regParam, and alpha.
1. Rank is the number of latent factors in the model.
2. MaxIter is the maximum number of iterations to run.
3. RegParam specifies the regularization parameter in ALS.
4. Alpha is a parameter applicable to the implicit feedback variant of ALS that governs the baseline confidence in preference observations.

In [ ]:
als = ALS(userCol="userId", itemCol="movieId", ratingCol="rating", coldStartStrategy="drop")

In [ ]:
#set the parameters and its values
paramGrid = (ParamGridBuilder()
                    .addGrid(als.rank, [ 8, 12])
                    .addGrid(als.maxIter, [5,10])
                    .addGrid(als.regParam, [0.01,0.001])
                    .addGrid(als.alpha, [1.0,2.0])
                    .build())
#evaluator is RMSE
evaluatorR = RegressionEvaluator(metricName="rmse", labelCol="rating")


# Run cross-validation, and choose the best set of parameters.
alsCV = CrossValidator(estimator=als,
                            estimatorParamMaps=paramGrid,
                            evaluator=evaluatorR,
                           numFolds=5)

In [ ]:
#fit the train dataset
modelCV_1 = alsCV.fit(train_1)

MLlib will automatically track trials in MLflow. After your tuning fit() call has completed, view the MLflow UI to see logged runs.

In [ ]:
# Best model and its parameters of Model 1 (60/40)
bestModel_1 = modelCV_1.bestModel
print ('Best Param (rank): ', bestModel_1._java_obj.parent().getRank())
print ('Best Param (MaxIter): ', bestModel_1._java_obj.parent().getMaxIter())
print ('Best Param (regParam): ', bestModel_1._java_obj.parent().getRegParam())
print ('Best Param (Alpha): ', bestModel_1._java_obj.parent().getAlpha())

Best Param (rank): 12
Best Param (MaxIter): 10
Best Param (regParam): 0.01
Best Param (Alpha): 1.0

In [ ]:
# Evaluate the best model of model 1 (60/40) by computing the RMSE on the test data
predictionsCV_1 = bestModel_1.transform(test_1)
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
rmseCV_1 = evaluator.evaluate(predictionsCV_1)
print("Root-mean-square error = " + str(rmseCV_1))

Root-mean-square error = 1.7826741552164505

In [ ]:
modelCV_2 = alsCV.fit(train_2)

MLlib will automatically track trials in MLflow. After your tuning fit() call has completed, view the MLflow UI to see logged runs.

In [ ]:
# Best model and its parameters of Model 2 (80/20)
bestModel_2 = modelCV_2.bestModel
print ('Best Param (rank): ', bestModel_2._java_obj.parent().getRank())
print ('Best Param (regParam): ', bestModel_2._java_obj.parent().getRegParam())
print ('Best Param (MaxIter): ', bestModel_2._java_obj.parent().getMaxIter())
print ('Best Param (Alpha): ', bestModel_2._java_obj.parent().getAlpha())

Best Param (rank): 12
Best Param (regParam): 0.01
Best Param (MaxIter): 10
Best Param (Alpha): 1.0

In [ ]:
# Evaluate the best model of Model 2 (80/20) by computing the RMSE on the test data
predictionsCV_2 = bestModel_2.transform(test_2)
evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating", predictionCol="prediction")
rmseCV_2 = evaluator.evaluate(predictionsCV_2)
print("Root-mean-square error = " + str(rmseCV_2))

Root-mean-square error = 1.6515568526657742

After tuning, the RMSE of model 1 (split by 60/40) decreased from 2.045 to 1.783. The RMSE of model 2 (split by 80/20) decreased from 1.982 to 1.652. 

Best Parameters: 
1. rank:  12
2. regParam:  0.01
3. maxIter:  10
4. alpha:  1.0

The values of maxiter and alpha remains the  default values, however, the values of rank change from 10 to 12 and the value of alpha changed from 1.0 to 0.01.

### 5. Calculate top 15 movies recommendations for user id 10 and user id 14. Show your code and output.

In [ ]:
# Generate top 15 movies recommendations for each user
userRecs = bestModel_2.recommendForAllUsers(15)

In [ ]:
#get top 15 movies recommendations for user id 10
userId10 = userRecs.where(col('userId')==10).select(col('recommendations'))
Top_10 = userId10.collect()
Top15_userid10 = []
for i in Top_10[0][0]:
    Top15_userid10.append(i[0])
#show the result of top 15 movies
print('Top 15 movies recommendations for user id 10 is:',Top15_userid10)

Top 15 movies recommendations for user id 10 is: [92, 2, 40, 89, 81, 25, 0, 42, 4, 49, 9, 59, 5, 87, 60]

In [ ]:
#get top 15 movies recommendations for user id 14
userId14 = userRecs.where(col('userId')==14).select(col('recommendations'))
Top_14 = userId14.collect()
Top15_userid14 = []
for i in Top_14[0][0]:
    Top15_userid14.append(i[0])
#show the result of top 15 movies
print('Top 15 movies recommendations for user id 14 is:',Top15_userid14)

Top 15 movies recommendations for user id 14 is: [52, 85, 63, 76, 72, 62, 96, 25, 90, 53, 29, 47, 91, 31, 58]